### PR TITLE
New Reduction Rule: 3-SAT to Independent Set Problem

### PR DESCRIPTION
**Pull request by SciCodePhy**

### CODE REVIEW

File: src/rules/sat_independentset.jl [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+function extract_solution(res::Reduction3SATToIndependentSet, sol)
+    if length( sol[1] ) >= res.k
+        return transform_is_to_3sat_solution(res.sat_source, sol, res.literal_to_nodes )
+    else
+        return Vector{Int}()
+    end
+end
+
+# ----Useful Functions----
+function reduce_3sat_to_independent_set(s::Satisfiability{T}) where T
```
Comment (GiggleLiu) [inline src/rules/sat_independentset.jl:27]: can we generalize this to generic sat?

File: src/rules/sat_independentset.jl [hunk 2]
```diff
@@ -0,0 +1,117 @@
+"""
+The reduction result of a 3-SAT problem to an Independent Set problem.
+"""
+struct Reduction3SATToIndependentSet{T, GT}
+    sat_source::Satisfiability{T}
+    is_target::IndependentSet{GT}
+    k::Int
+    literal_to_nodes::Dict{BoolVar{T}, Vector{Int}}
+end
+target_problem(res::Reduction3SATToIndependentSet) = res.is_target
```
Comment (GiggleLiu) [inline src/rules/sat_independentset.jl:10]: This line needs to be covered.

File: test/rules/sat_independentset.jl [hunk 3]
```diff
@@ -0,0 +1,24 @@
+using Test, ProblemReductions, Graphs
+
+@testset "sat_independentset" begin
+
```
Comment (GiggleLiu) [inline test/rules/sat_independentset.jl:4]: Needs more tests.

**Review by GiggleLiu — APPROVED** [review]
Good job.

File: src/rules/sat_independentset.jl (hunk 1 above)
Comment (SciCodePhy) [reply src/rules/sat_independentset.jl:27]: I will have a try! The common references only prove the reduction from 3-SAT to the Independent Set, but the mapping can be more general.

File: test/rules/sat_independentset.jl (hunk 3 above)
Comment (SciCodePhy) [reply test/rules/sat_independentset.jl:4]: Resolved: I add more tests.

File: src/rules/sat_independentset.jl (hunk 2 above)
Comment (SciCodePhy) [reply src/rules/sat_independentset.jl:10]: Resolved: Covered

**Comment by SciCodePhy** [comment]
I made a mistake in the first version:
When I extract solutions from the Independent Set to 3-SAT, I found that some solutions of Independent Set don't cover all the literals. Previously, I just threw away this type of solutions. However, this type of solutions correspond to the case when the uncovered literals can take either true of flase. Therefore, this type of solutions should be kept.

**Pull request merged**